### PR TITLE
feat(agents): declarative status rules for custom agents

### LINKED ISSUE #3137: feat: config-level status detection for custom agents (declarative rules / activity fallback / status CLI)
# feat: config-level status detection for custom agents (declarative rules / activity fallback / public status CLI)

_Note: this issue was drafted with an AI coding agent via back-and-forth. The investigation, measurements, and decisions are mine; most of the prose is the agent's._

## Summary

Custom agents (`[session.custom_agents]`) that are *similar to but not the same binary as* a built-in agent have no working path to status detection today. They sit at `Idle` forever, so status-transition notifications never fire for them. I'd like a config-level way to give them basic detection without a per-agent code PR, and I'm happy to implement whichever shape you prefer.

This is the narrower, concrete version of the open question in #3076 ("Could the custom-agent config grow a declarative status-rule form ... so the common case never needs a code PR?") — filed as its own feature request now that I've hit it in practice.

## What I ran into

I run a custom Claude-Code-like harness (`gjc`, its own config dir and its own hook format) under AoE:

```toml
[session.custom_agents]
gjc = "gjc"

[session.agent_detect_as]
gjc = "claude"
```

Measured on aoe 1.13.0, Linux:

| | without `agent_detect_as` | with `gjc = "claude"` |
|---|---|---|
| `AOE_INSTANCE_ID` injected | no | yes |
| status | Idle forever | Idle forever |

Both existing detection layers structurally can't work for this class of agent:

- **Pane parsing:** `agent_detect_as` aliases the built-in's detector, which matches the *built-in's* spinner/prompt strings, not the custom agent's output. Result is permanent Idle — or active misreads, as noted in #3076 (OMP aliased to `pi` reported "waiting" as idle because OMP has no `pi>` prompt).
- **Hooks:** `agent_status_hooks` installs hooks into the *built-in's* config file (`~/.claude/settings.json`, `~/.codex/hooks.json`). A harness with its own config/hook format never reads them. Hand-copying the hook commands into the custom agent's config does nothing for the same reason.

`agent_detect_as` is working as documented — it targets "same agent, launched differently" (the `lenovo-claude = "ssh -t lenovo claude"` case). The gap is the *similar-but-different harness* case, and it recurs with every new harness (gjc today, OMP in #3065 before #3073 made it a built-in, the next one tomorrow).

## Prior art in this repo

- #3076 concluded agents-as-plugins is too heavy, and that adding built-ins is fine "until it gets burdensome". This proposal is the config-level relief valve for the cases that don't clear the bar for a curated built-in.
- The plugin design (D7, #2096) already accepted **Tier 0 declarative ordered rules** — "Wildcard rules give custom `--cmd` agents basic detection" — but the status-detection contribution registry is deferred and not in the tree yet. This proposal is essentially shipping Tier 0 through session config now; if/when the plugin registry lands, it can subsume the same rule shape.
- `agents.<name>.status_map` covers hook-*event*-to-status remapping, which presupposes working hooks, so it doesn't reach this case.

### 1. Declarative status rules in config

```toml
[[session.agent_status_rules.gjc]]
status = "waiting"
contains = "(y/n)"

[[session.agent_status_rules.gjc]]
status = "running"
regex = "esc to interrupt|working|thinking"

# no match -> idle
```

Ordered rules, `contains`/`regex`, first match wins, evaluated on the same pane snapshot the built-in detectors already consume (`detect_status_from_content` path). Regexes compiled once at config load. An agent with rules gets them as its `detect_status`; everything else is untouched. This mirrors the shape most built-in detectors already have (per #3076, 13 of 16 are "a rule table wearing a function signature").

### 2. Generic activity fallback (complement or alternative)

Opt-in per agent: pane content changing → `running`, unchanged for N seconds → `idle`. Zero configuration and covers any future harness, at the cost of fidelity (no `waiting`, long non-streaming thinking reads as idle). Still strictly better than permanent Idle for notification purposes.

### 3. Public status-reporting CLI

```
aoe session set-status running|waiting|idle   # resolves session via $AOE_INSTANCE_ID
```

Formalizes the currently-private status-file contract (`/tmp/aoe-hooks-<euid>/$AOE_INSTANCE_ID/status`) so harnesses that have their own hook systems can report status directly and precisely. Smallest change, but only helps agents that expose hooks, so it complements rather than replaces 1/2. (`aoe status` already exists as the read-only summary; naming is yours.)

## Environment

- aoe 1.13.0 (also read against current `main`), Linux, tmux sessions (not structured view)
- claude and codex sessions on the same host detect correctly (for codex this needed the interactive hook trust approval — `t` — which was easy to miss; can send a small docs PR for that separately if wanted)

### PR DESCRIPTION
**Pull request by bjkim95**
## Description

Implements the primary proposal from #3137: `[[agents.<name>.status_rules]]` — declarative pane status rules that give a custom agent basic status detection without a code change.

A `[session.custom_agents]` harness that is *similar to but not the same binary as* any built-in currently has no working detection path: `agent_detect_as` aliases a built-in's pane detector (which matches the wrong output — cf. the OMP-aliased-to-`pi` misread noted in #3076), and `agent_status_hooks` installs hooks into the *built-in's* config file, which a custom harness never reads. Such sessions sit at Idle forever and never fire status-transition notifications.

With this PR:

```toml
[session.custom_agents]
gjc = "gjc"

[[agents.gjc.status_rules]]
status = "waiting"
contains = "(y/n)"

[[agents.gjc.status_rules]]
status = "running"
regex = "esc to interrupt|thinking"
```

Ordered rules, `status` + exactly one of `contains` (case-insensitive substring) or `regex` (Rust regex, as written), evaluated against the ANSI-stripped pane snapshot the built-in detectors already consume. First match wins; no match reports `idle`. This is the config-shaped subset of the plugin design's Tier 0 status rules (D7, #2096) — if/when the plugin status registry lands it can subsume the same rule shape.

Design notes:

- Rules compile once into a process-global registry (`tmux::status_rules`), installed by `profile_config::resolve_config` — the status poll hot path deliberately never loads config (same reason `Instance.detect_as` is resolved at build time). `aoe status` gets an explicit resolve since it previously loaded no config at all.
- Rules outrank an `agent_detect_as` alias (an agent with rules uses them even if an alias is configured) and a built-in detector of the same name (explicit user override).
- Malformed rules (no matcher, both matchers, invalid regex) are skipped with a `tracing` warning; an agent whose rules all fail to compile behaves as if none were configured.
- Hook-based detection is untouched: rules only participate in the pane fallback path, which is the only path a hookless custom agent ever reaches.

## PR Type

- [x] New Feature
- [ ] Bug Fix
- [ ] Refactor
- [ ] Documentation
- [ ] Infrastructure / CI

## Checklist

- [x] New and existing tests pass
- [x] Documentation was updated where necessary
- [ ] For UI changes: included screenshot or recording

## Test Coverage Analysis

**Web dashboard / structured view (Playwright user story)**
- [x] N/A: this PR doesn't change a user-facing dashboard flow

**TUI / CLI (e2e test)**
- [ ] N/A
- [ ] Added or updated an e2e test under `tests/e2e/`
- [x] Skipped a test, because: the new behavior is exercised end-to-end at the unit seam it lives behind — `detect_status_from_content` dispatch (rules vs built-in vs unknown-tool), registry install/replace, rule compilation edge cases, `detect_as` precedence, and config TOML round-trip are all covered in `src/tmux/status_rules.rs` and `src/session/config.rs` tests. An e2e run would additionally need a live tmux pane running a fake agent only to re-test the same dispatch line; happy to add one if you'd like.

Test details: `cargo test` on Linux — 4290 passed, 2 failed: `delete_branch_reaps_lingering_locked_worktree` and `purge_recovers_when_project_path_diverged_and_locked_entry_survives`, which fail identically on unmodified `origin/main` in my environment (git 2.34.1 worktree branch-delete behavior), i.e. pre-existing and unrelated. `cargo fmt --check` and `cargo clippy` clean.

## AI Usage

- [ ] No AI was used
- [x] AI was used

**AI Model/Tool used:**
Claude (agent harness) — investigation, implementation, and this PR text via back-and-forth; the problem report, measurements, design decisions, and review are mine.

**Any Additional AI Details you'd like to share:**
Same setup as issue #3137. I (the human) reviewed the full diff and ran the verification myself.

- [x] I am an AI Agent filling out this form (check box if true)

## Summary

- Added `agents.<name>.status_rules` for custom agent status detection.
- Added ordered, case-insensitive `contains` and Rust `regex` matchers.
- Applied rules to ANSI-stripped pane content, with `idle` as the fallback.
- Scoped rules to profiles and resolved them before status polling and session commands.
- Kept hook-based detection separate from pane-based rule detection.
- Added warnings for invalid rules and rules that shadow built-in detectors.
- Added documentation and tests for matching, precedence, profile isolation, dispatch, and TOML round-tripping.

## Benefits

Custom agents can report accurate statuses from pane output without changes to built-in detectors or hook-based detection. Profile isolation prevents configuration from affecting unrelated profiles.

## Potential enhancements

- Improve diagnostics for skipped rules and unmatched content.
- Add more matcher types or configurable regex options.
- Add broader integration coverage for profile-specific polling and session commands.

### CODE REVIEW

File: docs/guides/configuration.md [hunk 1]
```diff
@@ -191,12 +192,31 @@ agent_detect_as = { "lenovo-claude" = "claude" }
...
 - **`agent_acp_cmd`** (optional): ACP launch command that lets the agent run in the structured view (see below).
 - **`default_tool`** (optional): Can point at a custom-agent name to default new sessions to it.
 
 Custom agents are always shown as available in the picker since their command may target a remote host or wrapper. All three maps are editable in config files or the TUI settings screen and support profile/repo overrides; profile/repo values fully replace the global map (redeclare any agents you want to keep). The Web wizard can select a configured custom agent but does not expose or edit the command strings.
 
+#### Status rules for custom agents
+
+`agent_detect_as` only works when the custom agent renders the same output as the built-in it aliases. For a harness that is *similar to but not the same binary as* any built-in, declare pane status rules instead — no change to the agent needed:
```
[the remark this answers is not in the document]
Comment (bjkim95) [reply docs/guides/configuration.md]: Fixed in 813747c7, and swept the remaining em dashes out of the rest of the added prose (module docs, field docs, comments) per the AGENTS.md style rule.

File: src/session/instance.rs [hunk 2]
```diff
@@ -5048,11 +5048,9 @@ impl Instance {
...
         );
 
-        let detection_tool = if self.detect_as.is_empty() {
-            &self.tool
-        } else {
-            &self.detect_as
-        };
+        // The session's own configured status rules outrank an
+        // `agent_detect_as` alias; without rules the alias applies.
+        let detection_tool = tmux::status_rules::detection_tool(&self.tool, &self.detect_as);
```
[the remark this answers is not in the document]
Comment (bjkim95) [reply src/session/instance.rs]: Good catch, fixed in 813747c7. The identities are now split: the hook branch keeps the detect_as alias (`hook_tool`), since hooks are installed for and interpreted by the alias's reconcilers; the rules-over-alias preference (`pane_tool`) now applies only from the pane-fallback path down, including the Claude typed-prompt hold check. Updated the `detection_tool` helper doc to state it is pane-only.

File: src/session/profile_config.rs [hunk 3]
```diff
@@ -100,7 +100,13 @@ pub fn profile_has_overrides(config: &ProfileConfig) -> bool {
...
     let global = Config::load()?;
     let profile_config = load_profile_config(profile)?;
-    Ok(merge_configs(global, &profile_config))
+    let merged = merge_configs(global, &profile_config);
+    // (Re)install the declarative status-rule registry from the effective
+    // config. The status poll hot path never loads config, so this resolve
+    // — which every polling surface passes through at startup — is where
+    // `[[agents.<name>.status_rules]]` edits become visible.
+    crate::tmux::status_rules::install_from_config(&merged);
+    Ok(merged)
```
[the remark this answers is not in the document]
Comment (bjkim95) [reply src/session/profile_config.rs]: Fixed in 813747c7: the error branch of `resolve_config_or_warn` now installs the registry from the default config it returns, so a failed resolve can't leave rules from an earlier successful resolve active.

**Comment by Seluj78** [comment]
I approved the CI to run. You need a rebase and  I'll let @njbrake review :)

**Review by njbrake — CHANGES_REQUESTED** [review]
_Note: this review was drafted by Claude via back-and-forth with @njbrake. The reasoning and decisions are his; the prose is Claude's._

Thanks for this, and the docs in particular are good. The problem is real: a `[session.custom_agents]` harness that is not the same binary as a built-in has no working detection path today, `agent_detect_as` matches the wrong output, and `agent_status_hooks` writes into a config file the harness never reads. The config-shaped design is the right size for it, and I confirmed the change is a strict no-op for anyone who configures no rules: `detect()` starts with `reg.get(tool)?`, so an empty registry falls through to the built-in unchanged, and `detection_tool()` reduces to the pre-existing expression when `has_rules` is false.

I also checked the matcher itself and it is total and fail-safe. Every malformed `(contains, regex)` combination lands in a skip arm, and notably `regex = ""` is skipped by the `!pattern.is_empty()` guard rather than compiled into a match-all.

**The blocker is that the registry is process-global and profile-unaware, and `install_from_config` replaces the whole map.** `resolve_config` is called with many different profiles in one process, repeatedly, at runtime:

- `src/server/mod.rs:4136`, `reap_idle_sessions`, resolves config for each distinct profile of live sessions every 60 seconds. Two profiles with live sessions means the registry is rewritten twice per pass and ends up holding whichever profile resolved last, so the other profile's rules vanish for the next 60 seconds of polling.
- `src/hooks/targets.rs:136` loops every profile on disk.
- `src/server/acp_reconciler.rs:619` has the same shape.

Meanwhile the consumers are profile-agnostic: `poll_statuses_once` polls a snapshot containing instances from all profiles, and `detect_status_from_content(content, tool)` takes no profile at all. So even with stable install ordering, profile A's rules for agent `gjc` apply to profile B's `gjc` session. The net effect for a multi-profile user is time-dependent, cross-contaminated status detection, which also makes the doc line "Takes effect on the next config resolve" untrue in the general case.

Keying the registry on `(profile, agent)` and having install replace only that profile's entries fixes it. The smallest version of the call-site threading adds a sibling `detect_status_from_content_in(profile, ...)` so only `instance.rs` and `smart_rename.rs` change and `reconcile_waiting_hook` stays deliberately rules-free. Happy to sketch it if that helps.

**Two smaller things.**

`aoe ps` and `aoe session show` never install the registry before polling, so they report `Idle` for a rules-having custom agent that `aoe status` correctly reports as Running. `ps.rs` has no `resolve_config` call at all, and `load_instances` spans every profile when `--profile` is not explicit, which is the blocker above in miniature.

`rules_override_builtin_detector` installs rules for `"claude"` into the process-global registry, while seven `claude` tests in `status_detection.rs` assert through the same dispatcher on other threads in the same test binary. The module-local `Mutex` does not cover them. The window is short, so a clean local run does not rule it out, but this is the shape of flake that surfaces monthly in CI. Swapping the test to a built-in that no dispatcher-level test exercises is the cheap fix; `#[serial]` on all fifteen tests would also work but serializes tests that are currently parallel.

Two notes on things I checked and am **not** asking you to change, since both are the right call: a settings-schema entry is correctly not required, because `AgentRuntimeConfig` carries no `SettingsSection` and `validate_patch_with` rejects any patch whose top-level key has no descriptor, so `agents.*` is file and TUI trusted only and unreachable from a web PATCH (which also means an untrusted dashboard user cannot inject a regex the daemon compiles); and no migration is needed for an additive `#[serde(default)]` field.

Two corrections to note:

- The description says "Hook-based detection is untouched: rules only participate in the pane fallback path." `reconcile_waiting_hook` routes through `detect_status_from_content`, so rules for a built-in name do reach the hook path. Option (b) above makes the claim true rather than just documented.
- CodeRabbit's title-check warning about the `feat(agents):` scope is a false positive; `.github/workflows/pr-title-check.yml` sets `requireScope: false` and the repo's own history uses scoped prefixes.

One design point worth your consideration: because `detect()` returns `Some(Idle)` on no-match, a user who adds one rule to a built-in's `status_rules` to tweak a single state loses that agent's entire blocking-prompt and spinner detection and gets `Idle` for everything else. It is documented as an explicit override, but it deserves at least an install-time warning when rules shadow a built-in detector.

I have approved the workflow runs, so we will have real CI on this shortly. Nothing here was machine-verified before now, since the branch was rebased after the previous workflow approval.

**Comment by njbrake** [comment]
_Note: this comment was drafted by Claude via back-and-forth with @njbrake. The reasoning and decisions are his; the prose is Claude's._

Closing as stale. My change request from July 31 landed after your last push, and the branch now conflicts with main.

I'd still like this to land: the detection gap for custom agent harnesses is real and the config-shaped design is the right size for it. Address the review, rebase, and reopen (or ping me to reopen). Thanks, and the docs in this one were good.

**Comment by bjkim95** [comment]
_Note: drafted by Claude Code in back-and-forth with me. The design decisions are mine; the implementation and local verification were done by the agent under my direction._

Rebased onto main and addressed the July 31 review. What changed:

**Registry is now profile-scoped.** It is keyed on `(profile, agent)`, and `install_from_config(profile, ...)` replaces only that profile's entries instead of the whole map, so the per-profile `resolve_config` installs (idle reap, the hook-target loop, the acp reconciler) no longer clobber each other, and profile A's rules no longer apply to profile B's same-named agent. The entry points normalize the profile through `effective_profile`, so an empty `source_profile` keys the same slot as the resolved default on both install and lookup.

**The hook path is genuinely rules-free now.** `detect_status_from_content` is built-in-only; a new `detect_status_from_content_in(profile, ...)` is the rules-aware path used by the pane fallback and the detached rename. `reconcile_waiting_hook` calls the rules-free one, so "hooks are untouched by rules" is now true in code rather than just documented.

**Poll-only surfaces resolve config first.** `aoe ps`, `session show`, and `session capture` now resolve config for the relevant profile(s) before polling. While fixing `ps` I found its loader never stamped `source_profile` (it stayed empty), so even after resolving, the profile-keyed lookup would have missed for a profile-scoped rule; the loader now stamps it, with a regression test.

**Shadow warning.** `install_from_config` now warns when rules shadow a built-in detector, since no-match returns Idle and a single narrow rule on a built-in name would otherwise silently drop the rest of that agent's detection.

The flaky-test concern resolves structurally from the rules-free split: the `claude` dispatcher test in `status_detection.rs` no longer shares registry state with `rules_override_builtin_detector`.

Verified locally: `cargo fmt --check` and `cargo clippy -- -D warnings` clean; `cargo test --lib` green apart from two pre-existing env failures (`git::worktree::tests::delete_branch_reaps_lingering_locked_worktree`, `session::trash::tests::purge_recovers_when_project_path_diverged_and_locked_entry_survives`) that also fail on unmodified main here.

I don't have permission to reopen the PR myself (the button errors out), so @njbrake, could you reopen it when you have a chance? Thanks for keeping it alive, and for the detailed review.

**Review by njbrake — APPROVED** [review]
_Note: this review was drafted by Claude via back-and-forth with @njbrake. The reasoning and decisions are his; the prose is Claude's._

Approved after fixing the remaining non-default profile handling in `aoe status` and `session show`, correcting `session capture` status-rule precedence, rebasing onto current `main`, and preserving the new CityHall config overrides during conflict resolution. The full local Rust suite and every GitHub check pass.
